torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -579,13 +603,23 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
       auto& variable = replica.variables[intra_bucket_index];
       const auto offset = replica.offsets[intra_bucket_index];
       const auto length = replica.lengths[intra_bucket_index];
+
+      const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
+      bool global_unused =
+          local_used_maps_[var_idx.replica_index][var_idx.variable_index]
+              .item<int>() == 0;
```

Comment:
This tensor should be copied to CPU once and used many times. If this sits on GPU then this type of access will cause many GPU-to-CPU copies and add considerable overhead.

Reply:
Regarding the replica_index/variable_index: could you consider adding a vector of `VariableIndex` to the bucket replace so we don't need to mess with the grad accumulator pointer here? Would be a bit cleaner.

Reply:
> Regarding the replica_index/variable_index: could you consider adding a vector of `VariableIndex` to the bucket replace so we don't need to mess with the grad accumulator pointer here? Would be a bit cleaner.

Done as [043ea04](https://github.com/pytorch/pytorch/pull/28883/commits/043ea04c1de49ebf39186950513234a7e5bfd0c4)

Reply:
> This tensor should be copied to CPU once and used many times. If this sits on GPU then this type of access will cause many GPU-to-CPU copies and add considerable overhead.

Good point. Fixed as [https://github.com/pytorch/pytorch/pull/28883/commits/b12597171fb70f3d2c183b479876740c30d78f4a](b125971)